### DISCUSSION TITLE
Idea: Collaborative custody multisig co-signing module

### DISCUSSION THREAD

**Discussion by justinmoon** [fedimint/fedimint#1455] [Fedimint Modules]
Make a module that could act as a co-signer in a collaborative multisig -- for example, as a signer in Unchained Capital or Casa products.

**Comment by oleonardolima** [comment]
@justinmoon Do you have any resources you think would be helpful to research this topic? 

I got interested in it and I think it would be a good idea for SoB'23 :) 

@elsirion Do you think this fits in the "federate all the things" idea?

**Reply by elsirion** [reply]
Yes, definitely fits the description for SoB! Would be a bit hard to do for pre-taproot due to multi-party-ECDSA being gnarly. With taproot the federation could run FROST/ROAST and then participate with that signature in a traditional multisig.

For resources I'd refer to @justinmoon since he opened the discussion. I'd look into [PSBTs](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki), [Miniscript](https://bitcoin.sipa.be/miniscript/) and its [rust impl](https://docs.rs/miniscript/9.0.0/miniscript/) and [FROST](https://eprint.iacr.org/2020/852)/[ROAST](https://eprint.iacr.org/2022/550.pdf) ([great presentation](https://www.youtube.com/watch?v=f2soc95MWWY)).

**Reply by oleonardolima** [reply]
Good! Thanks for the resources, I'll look at those (it'll be pretty helpful) alongside Justin's suggestions.

**Comment by oleonardolima** [comment]
While researching about this topic I have found some interesting resources, for anyone also interested 😁 

- https://nunchuk.io/blog/the-missing-multisig-standard-rJDAu8d4s4Mg85ATpYrNJT
- https://nunchuk.io/blog/a-new-multisig-standard-and-nunchuk-10-pVDPJ4bAAg5v1bALEMzYXh
- https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018385.html
- https://github.com/bitcoin/bips/blob/master/bip-0129.mediawiki 
- https://unchained.com/labs/
- https://github.com/unchained-capital/caravan (might be useful for testing)

**Comment by m1sterc001guy** [comment]
I've been thinking about this as I also think it is an interesting use case. This might be easier to implement once Taproot multisig schemes like FROST/ROAST or MuSig2 that use schnorr signatures are more adopted? Otherwise the main issue I see is that the guardians of a federation need to produce an EDCSA signature on a PSBT for a multisignature wallet, but no guardian should have exclusive access to the private key. I think you could probably use multi-party EDCSA? Schnorr sigs would likely be easier though.

**Reply by oleonardolima** [reply]
Yes, you and eric raised the same point of attention hehe.

Agreed, I guess using Taproot schemes would be much easier (I definitely need to refresh my Taproot/Schnorr knowledge lol).

I'll comment here if anything interesting related to it arises after checking the resources Eric recommended.

**Comment by justinmoon** [comment]
At this point I would encourage you to just spend a little time learning how bitcoin's multisig works. Playing around with nunchuk or Sparrow wallet https://github.com/sparrowwallet/sparrow. You can try using these on testnet or signet so you don't need to use real money to play around.

You can also try just using `bitcoin-cli` to create a multisig PSBT. This doc explains the relevant commands https://github.com/bitcoin/bitcoin/blob/master/doc/psbt.md. You can try having 2 separate wallets in bitcoin core (`createwallet`) and pass the `-rpcwallet` command when signing the PSBTs to choose which wallet to sign with. I did some work with this stuff a couple years ago so I can answer questions. For these you can just use regtest locally.

I would say don't worry about building a fedimint module yet, just learn how multisig and PSBTs work. Then later on we can think about how to build a Fedimint module that does similar things.

**Reply by oleonardolima** [reply]
Great! Thanks for the suggestion 😁 

I'll check Sparrow/Nunchuk and play around with the bitcoin-cli to create a multisig psbt.

Awesome! I'll ping you if any questions arise then.
